### PR TITLE
Add metrics for sync result

### PR DESCRIPTION
**Pull request by sawsa307**
Created sync metrics collector to collect sync related metrics. Added metrics to collect the cumulative count of sync results. Specified metrics emit interval in flags.go.

### CODE REVIEW

File: pkg/neg/metrics/metrics.go [hunk 1]
```diff
@@ -82,6 +82,8 @@ var (
 			Subsystem: negControllerSubsystem,
 			Name:      negOpEndpointsKey,
 			Help:      "Number of Endpoints during an NEG Operation",
+			// custom buckets - [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, +Inf]
```
Comment (swetharepakula) [inline pkg/neg/metrics/metrics.go:85]: I think this commit can be pulled into its own PR and merged rather quickly. Can you pull this out?

Thanks!

Comment (sawsa307) [reply pkg/neg/metrics/metrics.go]: Removed. Thanks!

File: pkg/neg/metrics/neg_metrics_collector.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	)
+)
+
+type SyncerMetricsCollector interface {
+	UpdateSyncer(key negtypes.NegSyncerKey, err error)
+}
+
+type SyncerMetrics struct {
+	// countSinceLastExport tracks the count of errors occured since last export
+	countSinceLastExport map[syncError]int
```
Comment (sawsa307) [inline pkg/neg/metrics/neg_metrics_collector.go:52]: change it to syncResultCount

File: pkg/neg/metrics/neg_metrics_collector.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	// mu avoid race conditions and ensure correctness of metrics
+	mu sync.Mutex
+	// duration between metrics exports
+	metricsInterval time.Duration
+}
+
+// NewNEGMetricsCollector initializes SyncerMetrics and starts a go routine to compute and export metrics periodically.
+func NewNegMetricsCollector(exportInterval time.Duration) *SyncerMetrics {
+	return &SyncerMetrics{
+		countSinceLastExport: map[syncError]int{
```
Comment (swetharepakula) [inline pkg/neg/metrics/neg_metrics_collector.go:62]: Since this is a counter, we don't need to store this state. This will be necessary only when we have gauge metrics.

File: pkg/neg/metrics/neg_metrics_collector.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
[the remark this answers is not in the document]
Comment (sawsa307) [reply pkg/neg/metrics/neg_metrics_collector.go:2]: Updated.

File: pkg/neg/syncers/transaction.go [hunk 5]
```diff
@@ -217,7 +217,7 @@ func (s *transactionSyncer) syncInternalImpl() error {
 
 	currentMap, err := retrieveExistingZoneNetworkEndpointMap(s.NegSyncerKey.NegName, s.zoneGetter, s.cloud, s.NegSyncerKey.GetAPIVersion(), s.endpointsCalculator.Mode())
 	if err != nil {
-		return err
+		return fmt.Errorf("%w, reason: %v", metrics.ErrCurrentEPNotFound, err)
```
[the remark this answers is not in the document]
Comment (sawsa307) [reply pkg/neg/syncers/transaction.go:228]: Updated. Thanks!

File: pkg/neg/metrics/neg_metrics_collector.go (hunk 3 above)
Comment (sawsa307) [reply pkg/neg/metrics/neg_metrics_collector.go:62]: Updated. Thanks!

File: pkg/neg/syncers/transaction.go [hunk 6]
```diff
@@ -384,19 +389,19 @@ func (s *transactionSyncer) ensureNetworkEndpointGroups() error {
...
-// It returns false and the corresponding reason if one of the two checks fails:
+// checkValidEndpointInfo checks if endpoint information is correct.
+// It returns false with the corresponding reason if one of the two checks fails:
 //
 //  1. The endpoint count from endpointData doesn't equal to the one from endpointPodMap:
 //     endpiontPodMap removes the duplicated endpoints, and dupCount stores the number of duplicated it removed
 //     and we compare the endpoint counts with duplicates
 //  2. The endpoint count from endpointData or the one from endpointPodMap is 0
-func (s *transactionSyncer) isValidEndpointInfo(eds []negtypes.EndpointsData, endpointPodMap negtypes.EndpointPodMap, dupCount int) (bool, string) {
+func (s *transactionSyncer) checkValidEndpointInfo(eds []negtypes.EndpointsData, endpointPodMap negtypes.EndpointPodMap, dupCount int) (bool, *negtypes.NegSyncResult) {
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction.go:399]: nitL checkEndpointInfo or validateEndpointInfo

File: pkg/neg/syncers/transaction.go [hunk 7]
```diff
@@ -406,27 +411,35 @@ func (s *transactionSyncer) isValidEndpointInfo(eds []negtypes.EndpointsData, en
...
-	return true, negtypes.ResultSuccess
+	return true, negtypes.NewNegSyncResult(nil, negtypes.ResultSuccess)
 }
 
-// isValidEPField returns false and the corresponding reason if there is endpoint with missing zone or nodeName
-func (s *transactionSyncer) isValidEPField(err error) (bool, string) {
-	if errors.Is(err, ErrEPMissingNodeName) {
+// checkValidEPField checks if endpoints have valid field
+// It return the result boolean with the corresponding reason
+func (s *transactionSyncer) checkValidEPField(err error) (bool, *negtypes.NegSyncResult) {
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction.go:426]: nit: checkEPField or validateEPField

File: pkg/neg/syncers/transaction.go [hunk 8]
```diff
@@ -486,19 +499,38 @@ func (s *transactionSyncer) syncNetworkEndpoints(addEndpoints, removeEndpoints m
...
+		errorState := s.errorState
+		if errorState == negtypes.ResultInvalidAPIResponse {
+			syncResult = negtypes.NewNegSyncResult(negtypes.ErrInvalidAPIResponse, errorState)
+		}
+		if errorState == negtypes.ResultInvalidEPAttach {
+			syncResult = negtypes.NewNegSyncResult(negtypes.ErrInvalidEPAttach, errorState)
+		}
+		if errorState == negtypes.ResultInvalidEPDetach {
+			syncResult = negtypes.NewNegSyncResult(negtypes.ErrInvalidEPDetach, errorState)
+		}
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction.go:529]: you can use a switch and make the default case do the last one

File: pkg/neg/syncers/transaction.go [hunk 9]
```diff
@@ -486,19 +499,36 @@ func (s *transactionSyncer) syncNetworkEndpoints(addEndpoints, removeEndpoints m
...
 }
 
 // collectSyncResult collects the result of the sync and emits the metrics for sync result
 func (s *transactionSyncer) collectSyncResult(wg *sync.WaitGroup) {
 	wg.Wait()
+	s.syncLock.Lock()
+	defer s.syncLock.Unlock()
+
+	var syncResult *negtypes.NegSyncResult
+	if s.inErrorState() {
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction.go:519]: Do you need to check for error state here? We just need the syncResult, so there should be a nil case and the detach case added to the switch.

File: pkg/neg/syncers/transaction_test.go [hunk 10]
```diff
@@ -1983,20 +1983,20 @@ func TestIsValidEndpointInfo(t *testing.T) {
...
+			expectedResult: negtypes.NewNegSyncResult(nil, negtypes.ResultSuccess),
 		},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.desc, func(t *testing.T) {
-			if got, reason := transactionSyncer.isValidEndpointInfo(tc.endpointsData, tc.endpointPodMap, tc.dupCount); got != tc.expect && reason != tc.expectedReason {
-				t.Errorf("invalidEndpointInfo() = %t,  expected %t", got, tc.expect)
+			if got, result := transactionSyncer.CheckEndpointInfo(tc.endpointsData, tc.endpointPodMap, tc.dupCount); got != tc.expect || result.Result != tc.expectedResult.Result || !errors.Is(result.Error, tc.expectedResult.Error) {
+				t.Errorf("CheckEndpointInfo() = %t|%s|%t, expected %t|%s|%t", got, result.Result, result.Error, tc.expect, tc.expectedResult.Result, tc.expectedResult.Error)
```
Comment (swetharepakula) [inline pkg/neg/syncers/transaction_test.go:1993]: We should do `got` then `wanted` or `expected` based on go style guide.

Break it up a little more so that it is easier to understand the test and debug in the future:

```
got, result := transactionSyncer.CheckEndpointInfo(tc.endpointsData, tc.endpointPodMap, tc.dupCount)
if got != tc.expect {
    t.Errorf(...)
}
if result.Result != tc.expectedResult {
    t.Errorf(...)
}
```

same for the other test.

**Review by swetharepakula — COMMENTED** [review]
/lgtm
/approve
